plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,332 @@ extern const char* const state_history_plugin_abi;
-                state_history::state_request req;
-                fc::raw::unpack(ds, req);
-                std::visit( [self]( auto& r ) {
-                   self->process( r );
-                }, req );
-                self->start_read();
-             });
-          });
-   }
-
-   // should only be called once per session
-   std::string to_description_string() const {
+   std::string get_remote_endpoint_string() const {
       try {
-         boost::system::error_code ec;
-         auto re = socket_stream->next_layer().remote_endpoint(ec);
-         return boost::lexical_cast<std::string>(re);
+         if constexpr(std::is_same_v<SocketType, boost::asio::ip::tcp::socket>)
+            return boost::lexical_cast<std::string>(stream.next_layer().remote_endpoint());
+         return "UNIX socket";
       } catch (...) {
-         static uint32_t n = 0;
-         return "unknown " + std::to_string(++n);
+         return "(unknown)";
       }
    }
 
-   uint64_t get_log_entry_impl(const eosio::state_history::get_blocks_result& result,
-                               bool has_value,
-                               std::optional<state_history_log>& optional_log,
-                               std::optional<locked_decompress_stream>& buf) {
-      if (has_value) {
-         if( optional_log ) {
-            buf.emplace( optional_log->create_locked_decompress_stream() );
-            return std::visit([&](auto& r) { return optional_log->get_unpacked_entry( r.this_block->block_num, *buf ); }, result);
-         }
-      }
-      return 0;
-   }
-
-   uint64_t get_trace_log_entry(const eosio::state_history::get_blocks_result& result,
-                                std::optional<locked_decompress_stream>& buf) {
-      return std::visit([&](auto& r) { return get_log_entry_impl(r, r.traces.has_value(), plugin.get_trace_log(), buf); }, result);
-   }
-
-   uint64_t get_delta_log_entry(const eosio::state_history::get_blocks_result& result,
-                                std::optional<locked_decompress_stream>& buf) {
-      return std::visit([&](auto& r) { return get_log_entry_impl(r, r.deltas.has_value(), plugin.get_chain_state_log(), buf); }, result);
-   }
-
-   uint64_t get_finality_data_log_entry(const eosio::state_history::get_blocks_result_v1& result,
-                                        std::optional<locked_decompress_stream>& buf) {
-      return get_log_entry_impl(result, result.finality_data.has_value(), plugin.get_finality_data_log(), buf);
-   }
-
-   void process(state_history::get_status_request_v0&) {
-      fc_dlog(plugin.get_logger(), "received get_status_request_v0");
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<status_result_send_queue_entry<session>>(self);
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
+   void awake_if_idle() {
+      boost::asio::dispatch(strand, [this]() {
+         wake_timer.cancel_one();
```

Comment:
This is a quirky pattern.

asio & corountines are often demonstrated in use cases that I would call a sort of "1:1 synchronous usage", for example echoing back what was sent, or replying to an HTTP request that was received. In those examples, a single `async_receive` is always followed by a single `async_send`. This makes the coroutine usage simple and elegant, it ends up with something like
```c++
while(true) {
   co_await sock.async_receive(...);
   co_await sock.async_send(...);
}
```
But this is not the model of how state history's websocket protocol operates. It can stream a potentially unlimited `async_send`s from just a single `async_receive`, and during the streaming of these `async_send`s it might need to "pause" sending until some event from our main thread source (i.e. a new block is applied), and it still must service new `async_receive`s during the streaming of data. And, remember, both the "send side" and "receive side" need to be on a single strand.

There is only one example pattern of a similar communication protocol in asio documentation: the [coroutine based chat server](https://www.boost.org/doc/libs/1_85_0/doc/html/boost_asio/example/cpp20/coroutines/chat_server.cpp). This is where this quirky `steady_timer`-as-a-sort-of-condition-variable comes from.